### PR TITLE
feat(catalog): deterministic cross-server content_id

### PR DESCRIPTION
**Pull request by CoffeeKnyte**
Two feature commits: deterministic, cross-server `content_id` (a derivation core, generation wiring, a hot-query fast path, and a collision-safe value-remap migration), then an online re-ID that promotes untagged items to their deterministic id at first match.

Plus two follow-up commits from review: `fix(catalog): harden content_id parsing and merge per review` (the CodeRabbit findings — stricter anchored-id validation, frozen precedence, fail-closed canonicalize) and `refactor(contentid): URL-safe "-" separator in content_id` (see the separator note below).

**Scope:** id-system refactor (how `content_id` is minted/stored), not a new user-facing endpoint or feature — so it proceeds under the "bug fixes, refactors … proceed normally" carve-out of the v1 scope gate (`docs/architecture/v1-scope.md` is not yet locked). It lays groundwork for future cross-server share/sync, which will be filed as its own v1 capability proposal.

## Problem

**The same title on two servers was treated as two unrelated items.** Every logical movie, show, season, and episode was stamped with a per-server numeric id minted locally at scan time. Because that id is the anchor everything hangs off — artwork, metadata, watch history, progress, favorites, ratings, collections, credits — two servers holding the identical movie shared *none* of that state. Anyone operating more than one Silo, or trying to move/sync a profile's history between servers, saw the same film or show show up as a stranger on the other side: no resume point, no watched flags, no carried-over ratings.

**There was no stable identity to build cross-server features on.** Because the id was a time-ordered local counter, nothing about it could be reproduced on another machine, so any future "share/sync across servers" work had no ground truth to key on.

## Solution

The fix replaces the per-server counter with a structured natural key derived from the provider IDs already present in the library (`movie-tmdb-228064`, `series-tvdb-296762`, `episode-tvdb-296762-1-5`, `local-<hash>` fallback). Two servers that see the same tags mint the same id with no API call.

> **Separator note (post-review):** the component separator is `-`, not the `:` shown in earlier revisions. We switched it for readability/operability. `-` is an RFC 3986 *unreserved* character, so a `content_id` is URL-safe verbatim: it needs no percent-encoding and doubles as its own tidy path segment — `/item/series-tvdb-296762` — identical to the value stored in the database, instead of the `%3A`-escaped `/item/series%3Atvdb%3A296762` a colon produces. The stored value and the URL value are the same string, so there is no encode/decode boundary and an operator can grep an id straight out of a URL or log. Every component is `[a-z0-9]+` (or `tt`+digits), so `-` is an unambiguous delimiter. Structure and guarantees are unchanged.

#### Derivation core — `internal/contentid/contentid.go`

`ForMovie`/`ForSeries`/`ForSeason`/`ForEpisode`/`ForLocal`, the `SeriesIDFromContentID` string transform, and frozen provider precedence (movies `tmdb → imdb → tvdb`; series/season/episode `tvdb → tmdb → imdb`, matching where Radarr/Sonarr reliably tag). The load-bearing invariant: an episode/season id embeds its series anchor, so the show is a pure string transform of the episode id — no catalog lookup. `SchemeVersion = 1` freezes format + precedence as a code constant so the population is never mixed-version.

#### Generation wiring — `internal/metadata/service.go`

`deriveLogicalContentID`/`deriveSeasonContentID`/`deriveEpisodeContentID` replace `idgen.NextID()` at every mint site (skeleton creation, provider-match at `service.go:1499`, explicit + implicit seasons, episodes, scanner-fallback episodes). Tagged libraries get their deterministic id at scan time; untagged new items get a path-derived `local-` id. Convergence of same-title items on one server is preserved by the existing dedup (`findExistingByProviderIDs`) plus `UpsertTx`'s `ON CONFLICT (content_id) DO UPDATE` backstop.

#### Hot watch-history query — `internal/catalog/history_source.go`

`display_id` resolves an episode history row to its show via `seriesFromAnchoredEpisodeExpr` (pure `split_part` transform) for provider-anchored ids, probing only the `media_items` PK. Legacy/`local-` ids still fall through the `episodes` join, which is null-poisoned for anchored ids so the planner never descends `episodes_pkey` for them.

#### Migration (value remap) — `migrations/sql/20260612130000_deterministic_content_id.sql`

Collision-safe value remap (Sonyflake → structured key) across the dynamically-enumerated reference graph (65 columns) with `COLLATE "C"`, FK drop/recreate, trigger quarantine, a retained audit map, and a working `down`. Collisions and anchorless/non-movie-series items are detected and kept on Sonyflake — the migration never merges or drops a row. Approach chosen over a hand-listed column set so a missed soft-reference can't silently orphan rows; the value remap is self-protecting, so over-inclusion is safe.

### feat(catalog): re-ID untagged items to deterministic content_id at first match

Tagged libraries are born with the right id at scan time, but an *untagged* file gets a path-derived `local-` id and only learns its provider IDs later, when the match worker confirms a result. A single gate inside `mergeAndPersist` (`canonicalizeLocalContentID`) promotes that placeholder to its deterministic id at the moment of first confirmed match: if the deterministic id is free the row is renamed in place (the DB moves every child row via `ON UPDATE CASCADE`, so it is a handful of rows for a fresh skeleton, not the full-table remap the migration does); if a row already sits at that id the two are the same logical item and the existing merge path (`rebindItemToExistingItem`) folds them together. The guard is a single `IsLocal` prefix check, so tagged content and all refreshes pay nothing, and the promotion is self-healing — a partial run simply re-runs on the next match.

```
SCAN a new file                          ← unchanged, still no network call
      │
      ▼
createOrFindSkeleton ──► tag?  YES ► movie-tmdb-123   NO ► local-<hash>
      │
      ▼
(later) match worker → mergeAndPersist
      │
      ▼
┌─────────────────────────────────────────────────────────┐
│ canonicalizeLocalContentID                               │
│                                                          │
│  Is the id still local- AND do we now have provider IDs? │
│        │ no  → do nothing, carry on                      │
│        │ yes                                             │
│        ▼                                                 │
│  Does a row already sit at movie-tmdb-123 ?              │
│        ├─ yes → MERGE this row into it  (existing code)  │
│        └─ no  → RENAME local-<hash> → movie-tmdb-123     │
│                   │                                       │
│                   ▼                                       │
│           DB auto-moves all child rows (ON UPDATE CASCADE)│
└─────────────────────────────────────────────────────────┘
      │
      ▼
rest of mergeAndPersist runs with the corrected id
(title, artwork, cast, episodes — all written under movie-tmdb-123)
```

The rename primitive (`silo_rename_content_id`, in `internal/metadata/canonicalize.go`) reuses the migration's catalog-driven column enumeration so the two stay in lockstep, and is added alongside the `ON UPDATE CASCADE` clauses in `migrations/sql/20260614120000_content_id_online_reid.sql`. ON UPDATE CASCADE only fires when a `content_id` is updated (essentially never outside this promotion), so there is no steady-state read/write cost; the one-time cost is the FK drop/recreate during the migration.

## Performance — measured at production scale

Measured on an isolated, prod-tuned PostgreSQL 18 instance (`shared_buffers=4GB`) loaded with an **exact-cardinality copy of production** (`cprod-postgres`): 184k items, 1.93M episodes, 775k watch-history rows / 162 profiles. "Before" reproduces production today (Sonyflake numeric `content_id`, `en_US.utf8`); "after" applies this PR's structured-key remap + `COLLATE "C"`. The 100-client load test ran on the isolated instance, never against production. **Every metric improves — this is a speedup, not a downgrade.**

**1. `content_id` index-probe cost** — 300k forced point lookups over 1.9M keys (the operation the hot paths are built from):

The last three columns are the point: speed alone wasn't the goal, so the two *faster* schemes were rejected for giving up properties the feature exists for. ✓ = has the property, ✗ = lacks it.

| Key scheme | Index size | µs / probe | vs today | Cross-server deterministic | Zero-join show transform | Human-readable |
| ---------- | ---------- | ---------- | -------- | :------------------------: | :----------------------: | :------------: |
| Sonyflake text, `en_US.utf8` *(today)* | 74 MB | **5.05** | 1.00× | ✗ | ✗ | ✗ |
| 128-bit hash `bytea(16)` | 74 MB | 1.80 | 2.81× | ✓ | ✗ | ✗ |
| `bigint` surrogate (≈ old Sonyflake) | 41 MB | 1.23 | 4.11× | ✗ | ✗ | ✗ |
| **Structured `content_id` + `COLLATE "C"` (this PR)** | 84 MB | **1.86** | **2.71×** | ✓ | ✓ | ✓ |

The locale-aware `strcoll` on today's text column is the bottleneck; `COLLATE "C"` (byte-wise `memcmp`) removes it. The structured key lands within **3%** of a 128-bit hash and within **1.5×** of a raw `bigint`, and is the **only all-✓ row**:

- **128-bit hash `bytea(16)`** — cross-server deterministic (it hashes the same natural key), but an opaque digest can't be string-transformed, so episode→series needs a catalog join (no zero-join hot path), and it isn't legible in logs, URLs, or `psql`.
- **`bigint` surrogate** — fast only because it's 8 bytes; it's the status-quo serial: per-server (not reproducible elsewhere), no embedded anchor to transform, not meaningful to a human.

Index grows ~14% (84 vs 74 MB), immaterial.

**2. Real history-page query** — heaviest profile (25,384 rows), from `buildHistoryDisplayBaseQuery`, warmed median of 6:

| State | Latency | `episodes_pkey` probes | Speedup |
| ----- | ------- | ---------------------- | ------- |
| Before (Sonyflake + `en_US.utf8`) | 385 ms | 25,348 | 1.00× |
| **This PR (structured + `COLLATE "C"`)** | **150 ms** | **2,775** | **2.57×** |

`COLLATE "C"` removes the locale-aware comparison cost, and structured keys let anchored episode ids skip the `episodes_pkey` descent (25,348 → 2,775 probes, the M1 fix).

**3. Concurrency** — `pgbench`, workload randomized across 150 real profiles (≥200 history rows each), 20 s per level:

| Concurrency | Before (lat / tps) | This PR (lat / tps) | Throughput gain |
| ----------- | ------------------ | ------------------- | --------------- |
| 1   | 43.8 ms / 23  | 13.0 ms / 77  | 3.4× |
| 10  | 50.9 ms / 196 | 21.1 ms / 473 | 2.4× |
| 50  | 152 ms / 329  | 91 ms / 548   | 1.7× |
| **100** | **311 ms / 321** | **181 ms / 551** | **1.7×** |

Comparison cost is pure CPU, so it dominates under concurrency: the cheaper `COLLATE "C"` comparisons raise the throughput ceiling from ~321 to ~551 tps and cut tail latency at 100 concurrent users by 42%. *(Warmed — dataset fits in `shared_buffers`; isolated instance shares a host with production, so the before/after ratios are the robust result, not absolute tps.)*

## Risk / follow-ups

- Re-ID-on-match for untagged-then-matched items now runs inline at first confirmed match (see the Re-ID subsection above). The one remaining case is a series that accumulated season/episode rows *before* it matched: those children keep their Sonyflake ids until a follow-up `recomposeSeriesChildIDs` pass re-derives them. At first match the children usually do not exist yet, so this is rare; it is not yet wired. Promotion also changes a primary key, so an item's anchored id can vary during its first-match window (accepted — no alias/redirect table).
- Cross-provider reconciliation (a series tagged tmdb-only on one server, tvdb-only on another) is not yet handled — both paths derive from the denormalized provider columns; the proper fix consults `media_item_provider_ids`.
- The migration runs in one transaction holding `AccessExclusive` locks for the COLLATE rewrite and remap — run off-peak; very large installs (10–50M rows) should use the batched procedure noted in the migration header.
- `local-` ids hash the absolute path, so a same-server remount re-IDs local items (acceptable for the best-effort local namespace).
- **Client coordination.** `content_id` stays a `string` in every API response (no additive-only violation), but the value-remap migration changes the *values* clients use as their anchor. Any `silo-android` / `silo-apple` state keyed on the old Sonyflake ids (cached offline downloads, stored references) is stale after migration and needs a re-fetch/invalidation path — tracked as client follow-up, not in this PR.

## Verification

- Unit tests: `internal/contentid` (derivation, precedence, the composition↔transform round-trip invariant, edge cases) pass; `internal/catalog` and `internal/metadata` suites pass with updated history-query golden fragments.
- `go build ./internal/...`, `go vet`, `gofmt -l` all clean.
- Migration on real schema (PG18 + pgvector): full migration chain plus this one applies, reverts (`down`), and re-applies (`up`) cleanly — recollates 65 columns, drops/recreates 27 FKs and all `episode_catalog` triggers, retains only the audit map.
- Online re-ID migration (`20260614120000_content_id_online_reid`): applies on the real schema after the value-remap migration; every content_id-family FK gains `ON UPDATE CASCADE` while keeping its original `ON DELETE` action (e.g. `episodes_series_id_fkey` becomes `ON UPDATE CASCADE ON DELETE CASCADE`); `make migrate-validate` passes; full `up`/`down`/`up` cycle is clean (down drops `silo_rename_content_id` and strips only `ON UPDATE CASCADE`).
- Online re-ID functional test: `silo_rename_content_id` on a local series moves the `media_items` PK, cascades `seasons.series_id`/`episodes.series_id`, sweeps the provider-id soft references, and correctly leaves the children's own (Sonyflake) ids untouched (the deferred `recomposeSeriesChildIDs` case); a local movie rename moves the item and its provider ids; `ON DELETE CASCADE` still fires (delete leaves 0 orphans).
- Migration data-bearing test: correct precedence; soft-reference + FK-child remap; 0 FK orphans; collisions/unmatched correctly stay on Sonyflake; the collision cascade keeps a unique episode of a colliding series off an orphaned anchor; `down` reverts values and collation.
- Performance: benchmarked against an exact-cardinality copy of `cprod-postgres` (1.93M episodes, 775k history rows) — 2.57× faster history page, 1.7× throughput at 100 concurrent users, 2.7× cheaper per `content_id` probe (see Performance section).

## AI-use disclosure

Implemented with AI assistance (Claude). An independent adversarial review (Fable 5) was run against the spec's required properties; its critical/major findings were fixed and re-validated, with the larger items recorded as the follow-ups above. Performance benchmarking was AI-assisted against an isolated copy of production data.

## Summary by CodeRabbit

* **New Features**
  * Added deterministic, provider-anchored identifiers for movies, series, seasons, and episodes to keep items consistent across server instances.
  * Canonicalizes local placeholder identifiers to deterministic provider identifiers during metadata ingestion.
  * Added reversible ID encoding support for compatibility item/season IDs.
* **Bug Fixes**
  * Improved watch/history resolution for provider-anchored episode content so related items link correctly.
* **Documentation**
  * Added a developer reference describing the deterministic identifier format and migration strategy.
* **Tests**
  * Expanded coverage for deterministic ID composition/parsing/packing, codec behavior, history query generation, and migration timeout handling.
* **Chores**
  * Added migrations to remap existing `content_id` values and enabled online re-ID with cascading updates; migration timeout is now configurable via an environment variable.

### CODE REVIEW

File: internal/catalog/history_source.go [hunk 1]
```diff
@@ -206,16 +206,47 @@ func buildHistoryDisplayBaseQuery(access AccessFilter, snapshot *time.Time) (str
...
 		FROM (
-			SELECT COALESCE(NULLIF(e.series_id, ''), h.media_item_id) AS display_id, h.watched_at
+			SELECT %[1]s AS display_id, h.watched_at
 			FROM user_watch_history h
-			LEFT JOIN episodes e ON e.content_id = h.media_item_id
-			JOIN media_items mi ON mi.content_id = COALESCE(NULLIF(e.series_id, ''), h.media_item_id)
-			WHERE %s
+			LEFT JOIN episodes e
+				ON e.content_id = CASE WHEN h.media_item_id LIKE 'episode:%%' THEN NULL ELSE h.media_item_id END
+			JOIN media_items mi ON mi.content_id = %[1]s
```
[the remark this answers is not in the document]
Comment (CoffeeKnyte) [reply internal/catalog/history_source.go:238]: Fixed in `eecfe7c4`. Extracted a shared `anchoredEpisodePredicate` that gates **both** the null-poisoned episodes join key and the series-recovery transform on the full five-part shape (`split_part` parts 2–5 all non-empty), not just the kind prefix. A malformed value like `episode-broken` no longer transforms to `series-broken-` and vanish at the `media_items` join — it fails the predicate and falls through to the `episodes` lookup. (The separator is now `-` per the post-review note, so the predicate matches `episode-%` / `split_part(..., '-', N)`.)

File: internal/contentid/contentid.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+// version, which provider anchors a logical item when more than one tag is
+// present. Two servers that both see the tags therefore pick the same anchor.
+//
+//   - Movies prefer TMDB (the richest movie source).
+//   - Series/Season/Episode prefer TVDB (the traditional episode-canonical
+//     source).
+var (
+	MovieProviderPrecedence  = []string{ProviderTMDB, ProviderIMDB, ProviderTVDB}
+	SeriesProviderPrecedence = []string{ProviderTVDB, ProviderTMDB, ProviderIMDB}
+)
```
[the remark this answers is not in the document]
Comment (CoffeeKnyte) [reply internal/contentid/contentid.go:94]: Fixed in `eecfe7c4`. `MovieProviderPrecedence`/`SeriesProviderPrecedence` are now unexported (`movieProviderPrecedence`/`seriesProviderPrecedence`) and never returned by reference, so no package can reorder the frozen `SchemeVersion` precedence at runtime. Confirmed zero external callers before unexporting.

File: internal/contentid/contentid.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+		// remaining components), re-prefix with "series:".
+		rest := id[strings.IndexByte(id, ':')+1:]
+		parts := strings.SplitN(rest, ":", 3)
+		if len(parts) < 2 {
+			return "", false
+		}
+		if _, valid := normalizeProviderID(parts[0], parts[1]); !valid {
+			return "", false
+		}
+		return kindSeries + ":" + parts[0] + ":" + parts[1], true
```
[the remark this answers is not in the document]
Comment (CoffeeKnyte) [reply internal/contentid/contentid.go]: Fixed in `eecfe7c4`. Added `parseAnchored`, which enforces the exact per-kind arity and requires the trailing season/episode components to be base-10 integers. `SeriesIDFromContentID` and `IsProviderAnchored` both route through it, so truncated forms like `episode-tvdb-296762` (missing season/episode) and `season-tvdb-296762` (missing season number), plus non-numeric suffixes, now fail closed instead of masquerading as valid anchors. Regression cases added to the test table.

File: internal/metadata/canonicalize.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+		return "", err
+	}
+	if !contentid.IsProviderAnchored(target) || target == from {
+		return from, nil
+	}
+
+	// A row already at the target id is the same logical item; merge onto it.
+	if existing, err := s.itemRepo.GetByID(ctx, target); err == nil && existing != nil {
+		if err := s.rebindItemToExistingItem(ctx, from, target, false); err != nil {
+			return "", fmt.Errorf("merging local item %s into %s: %w", from, target, err)
```
[the remark this answers is not in the document]
Comment (CoffeeKnyte) [reply internal/metadata/canonicalize.go]: Fixed in `eecfe7c4`. `canonicalizeLocalContentID` now distinguishes `catalog.ErrItemNotFound` (target genuinely free → rename) from a real lookup error (returned as a retryable error) — a transient failure no longer falls through into the rename path and surfaces as a misleading unique-constraint conflict. I also flipped `allowMatchedSource` to `true`, so a refresh of an already-matched `local-` source is consolidated onto the canonical row instead of orphaning a duplicate; this runs under the provider-dedup lock the caller already holds.

File: migrations/sql/20260612130000_deterministic_content_id.sql [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+-- unrelated data; it only needs the names to genuinely hold content ids.
+INSERT INTO _cid_cols
+SELECT c.oid::regclass, a.attname
+FROM pg_class c
+JOIN pg_namespace n ON n.oid = c.relnamespace
+JOIN pg_attribute a ON a.attrelid = c.oid AND a.attnum > 0 AND NOT a.attisdropped
+JOIN pg_type t ON t.oid = a.atttypid
+WHERE c.relkind IN ('r', 'p')
+  AND n.nspname = 'public'
+  AND t.typname IN ('text', 'varchar', 'bpchar')
```
[the remark this answers is not in the document]
Comment (CoffeeKnyte) [reply migrations/sql/20260612130000_deterministic_content_id.sql:170]: Not changed — verified not applicable here. I grepped every content-id-holding column declaration in `migrations/sql/`, and they are **all `text`**; none are `varchar`/`bpchar`. The `varchar`/`bpchar` entries in the `_cid_cols` type filter are deliberate defensive breadth so the sweep cannot miss a soft reference — but because no such column actually exists in the schema, `ALTER ... TYPE text` is a no-op for the column *type* in practice, and the `down` migration symmetrically restores `text COLLATE pg_catalog."default"`.

The suggested fix is also incomplete for the case it targets: capturing `typname` alone would not preserve a `varchar(N)` length modifier — doing it correctly would require `format_type(atttypid, atttypmod)` and a matching `down`. Since there are zero non-`text` columns to preserve, that added complexity is not warranted today.

I left a marker for the future: if a non-`text` content-id column is ever introduced, this should be revisited (capture `format_type(...)` in `_cid_cols` and reuse it in both `ALTER` loops). Happy to add a `format_type`-based version now if you would rather have it be defensive regardless.

**Comment by CoffeeKnyte** [comment]
### Why this scheme generalizes (movies/shows today, more later)

A note on where this can go, since it's built to extend rather than be reworked:

**Today it covers movies and shows well.** Movies anchor on `tmdb → imdb → tvdb`, series/season/episode on `tvdb → tmdb → imdb` — the providers Radarr/Sonarr/Plex already tag reliably — and an episode/season key embeds its series anchor, so the show falls out by pure string transform (the zero-join history path).

**New entity types are additive.** The leading `kind` token (`movie`/`series`/`season`/`episode`/`local`) domain-separates the namespaces, so adding `audiobook-…`, `ebook-…`, `podcast-…`, `music-…` down the line is just a new kind token plus a precedence list for that type's providers (Audible, Google Books, MusicBrainz, …). Those types stay on Sonyflake for now; promoting any of them is a self-contained follow-up that touches no existing keyspace.

**A new default metadata provider is also additive.** Provider precedence is a frozen, ordered list under `SchemeVersion`. Introducing a new source — say **MetaDB** in future — and making it the default anchor is just prepending it to the relevant precedence list (plus its id validation in `normalizeProviderID`) and bumping `SchemeVersion`. That version bump is exactly what the value-remap migration machinery in this PR is built for: it re-derives every affected id, detects collisions, keeps un-derivable rows on their old id, and retains an audit map — so switching the default is a controlled, reversible remap rather than a flag day. Two servers that both adopt the new default still converge on the same id.

In short, the structure itself — `kind-provider-id`, frozen precedence, version-gated remap — is the extension point: broaden coverage by adding tokens/providers, not by reworking the anchor.

**Review by Quick104 — COMMENTED** [review]
Review focused on the seams where remapping *the* universal anchor leaks into systems that consume `content_id`. The architecture is sound — provider-anchored natural key is the right model and matches Plex/Jellyfin prior art, the string-transform invariant and `COLLATE "C"` analysis hold up, and the prior CodeRabbit findings are addressed. Three issues below that need cross-package reasoning the bots don't do. Two have inline comments; the first can't (its file isn't in this diff).

---

### 1. (Major — robustness) jellycompat item-UUID decode regresses from stateless to in-memory-cache-dependent across restarts

`internal/jellycompat/idcodec.go:82` `EncodeStringID` branches on whether the `content_id` parses as a `uint64`:

- **Before this PR** content_ids were numeric Sonyflakes → they take the `EncodeNumericID` branch, so the client UUID is a pure bit-packing and `DecodeID` (`idcodec.go:163`) reverses it with **zero state**. Survives restarts.
- **After this PR** content_ids are structured strings (`movie-tmdb-228064`) → they take the `uuid.NewSHA1` branch and reversibility depends **entirely on the in-memory `c.reverse` map** (`idcodec.go:91-95`), which is empty after a process restart and only repopulated when an item is re-encoded during the current lifetime.

The decode sites are by-id URL fetches — `DecodeStringID(EncodedIDItem, chi.URLParam(r, "id"))` at `handlers_items.go:441/645/1019/1114`, plus images (`images.go:96`) and userdata (`handlers_userdata.go:206`). The codec is a process-lifetime singleton (`router.go:261`). So after a server restart, a Jellyfin client (JellyCon/Findroid/Infuse) that presents a previously-issued item/season UUID — resume-from-home, a deep link, a detail page, an image request — hits a cold map, the stateless decode fails, and it returns `unknown compat id`; the item effectively 404s until something re-lists it.

This is exactly the failure path the repo lists as a core priority ("predictable under load and during failures — session restarts, reconnects"), and the PR's disclosure understates it: it says caches "reset once," but this is an *ongoing* reversibility dependency on warm in-process state, not a one-time migration cost.

SHA1 isn't invertible and a full scan is O(n), so a real fix needs either a **persisted `uuid → content_id` map** (write-through on encode + lookup on decode-miss) or **warming the map at startup** from the catalog. Both have real tradeoffs — added DB write on the id-encode hot path, or boot latency + memory for millions of episode/season entries — so this deserves its own focused, benchmarked change rather than a drive-by edit to a shared hot-path codec. Not blocking the concept; should be resolved or explicitly tracked as a jellycompat follow-up before merge.

File: migrations/sql/20260612130000_deterministic_content_id.sql [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
```
Comment (Quick104) [inline migrations/sql/20260612130000_deterministic_content_id.sql:170]: **(Major — data integrity) Array-valued soft reference `trending_discover_snapshots.content_ids` is silently skipped by the remap.**

The migration's central safety claim is that the catalog-driven sweep is self-protecting and "a missed soft-reference can't silently orphan rows." Here's a concrete counterexample: `trending_discover_snapshots.content_ids` (`migrations/sql/167_...:11`) is a `text[]` of resolved library content IDs. It's excluded from this sweep on **both** axes — the type filter on this line is scalar (`text`/`varchar`/`bpchar`, so arrays don't match) and the name allowlist below has `content_id`, not the plural `content_ids` — and an array can't carry an FK, so nothing else touches it. After the remap those arrays hold stale Sonyflake ids that resolve to nothing.

Severity is bounded: `internal/sections/trending_snapshot.go` refreshes the table via `INSERT … ON CONFLICT DO UPDATE`, so it self-heals on the next snapshot refresh — trending/discover rows are just broken in the interim. But the "cannot orphan" invariant has a real hole. I verified collection membership (`library_collection_items`, `user_personal_collection_items`, `abs_collection_items`) is FK-constrained and correctly caught; this `text[]` is the only column that slips through.

Element-wise remap (preserving order, leaving unmapped/collision elements untouched) closes it. I'm including this fix in a follow-up commit on this branch, in both this migration and `silo_rename_content_id`.

File: internal/metadata/canonicalize.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	if err := s.renameContentID(ctx, from, target); err != nil {
+		return "", err
+	}
+	return target, nil
+}
+
+// renameContentID moves a single content_id value (a media item, season or
+// episode) to a new, currently-free id. FK children follow via ON UPDATE
+// CASCADE; the silo_rename_content_id function also sweeps the unconstrained
+// soft references. The function body runs as one statement, so the move is
```
Comment (Quick104) [inline internal/metadata/canonicalize.go:96]: **(Minor — doc accuracy)** This comment is correct: `silo_rename_content_id`'s body does run as one statement, so the *rename* is atomic. The nit is with the PR description, which generalizes atomicity to the whole local→canonical promotion. The rename runs on `s.dbPool` here, while the rest of `mergeAndPersist` is separate work on the same pool — they aren't one transaction. The safety on partial failure actually comes from idempotent retry: if the process dies after the rename, the next match sees a non-local `contentID`, the `IsLocal` guard skips canonicalization, and the merge proceeds on the canonical id. That's genuinely self-healing and fine — it's just "retry-safe," not "atomic." Suggest tightening the PR-body wording; no code change needed.

**Comment by Quick104** [comment]
Pushed two follow-up commits to this branch addressing the review:

- **`b5d93b4` — finding #2 (array-valued soft reference):** the value-remap migration now remaps `trending_discover_snapshots.content_ids` element-wise (Up and Down), and `silo_rename_content_id` gets the matching `array_replace` so the runtime path stays in lockstep. Verified on PG18: mixed/collision/empty arrays remap correctly, order preserved, empty stays `{}` (no NOT NULL collapse), Down is an exact inverse.

- **`a44a301` — finding #1 (restart-decode regression):** resolved by making the compat encoding *reversible* instead of stateful, rather than adding a map. `internal/contentid` gains `Pack`/`Unpack` (a bit-packed ≤15-byte form of a structured/local content_id), `ForLocal` shrinks to a 112-bit hash so it packs losslessly, and `jellycompat` packs item/season ids into the UUID directly — decode is now pure computation, stable across restarts *and* instances, no lookup table. Numeric ids and arbitrary names (genres/studios) are unchanged. TDD'd with a cross-instance decode test that fails on the old hash+map path.

Finding #3 (atomicity wording) is a doc-only nit — no code change; flagged inline for the PR description.

AI-use disclosure: implemented with Claude (Opus 4.8).
